packages/flame/lib/src/sprite_batch.dart

```diff
@@ -144,37 +148,99 @@ class SpriteBatch {
 
   FlippedAtlasStatus _flippedAtlasStatus = FlippedAtlasStatus.none;
 
-  /// List of all the existing batch items.
-  final _batchItems = <BatchItem>[];
+  /// Stack of available (freed) indices using ListQueue as a stack.
+  final Queue<int> _freeIndices = Queue<int>();
 
-  /// The sources to use on the [atlas].
-  final _sources = <Rect>[];
+  /// Returns the total number of indices that have been allocated.
+  int get allocatedCount => _nextIndex;
+
+  /// Returns the number of currently free indices.
+  int get freeCount => _freeIndices.length;
+
+  /// The next index to allocate if no free indices are available.
+  int _nextIndex = 0;
+
+  /// A map to keep track of the logical index of each batch item by its id.
+  final Map<String, int> _idToIndex = {};
+
+  /// Returns all current ids
+  Iterable<String> get ids => _idToIndex.keys;
+
+  /// Sparse array of batch items, indexed by allocated indices.
+  final Map<int, BatchItem> _batchItems = {};
 
-  /// The sources list shouldn't be modified directly, that is why an
-  /// [UnmodifiableListView] is used. If you want to add sources use the
-  /// [add] or [addTransform] method.
-  UnmodifiableListView<Rect> get sources {
-    return UnmodifiableListView<Rect>(_sources);
+  /// Returns the number of indices currently in use.
+  int get usedCount => _nextIndex - _freeIndices.length;
+
+  /// Allocates a new index, reusing freed indices when possible.
+  int _allocateIndex() {
+    if (_freeIndices.isNotEmpty) {
+      return _freeIndices.removeFirst();
+    }
+    return _nextIndex++;
+  }
+
+  /// Frees an index to be reused later.
+  void _freeIndex(int index) {
+    _freeIndices.addFirst(index);
   }
 
+  /// The sources to use on the [atlas].
+  final Map<int, Rect> _sources = {};
+
   /// The transforms that should be applied on the [_sources].
-  final _transforms = <RSTransform>[];
+  final Map<int, RSTransform> _transforms = {};
 
-  /// The transforms list shouldn't be modified directly, that is why an
-  /// [UnmodifiableListView] is used. If you want to add transforms use the
-  /// [add] or [addTransform] method.
-  UnmodifiableListView<RSTransform> get transforms {
-    return UnmodifiableListView<RSTransform>(_transforms);
+  /// The colors to use for each batch item.
+  final Map<int, Color> _colors = {};
+
+  /// Whether the render lists are dirty and need to be rebuilt.
+  bool _dirty = true;
+
+  /// The lists used for rendering the batch items.
+  final List<Rect> _sourcesList = [];
+
+  /// The transforms used for rendering the batch items.
+  final List<RSTransform> _transformsList = [];
+
+  /// The colors used for rendering the batch items.
+  final List<Color> _colorsList = [];
```

Comment:
Why are the `BatchItem`s not responsible for their own properties? :thinking: 
It feels a quite messy having them here, and I don't see the performance gain?